call.go

```diff
@@ -113,6 +133,22 @@ func Invoke(ctx context.Context, method string, args, reply interface{}, cc *Cli
 			o.after(&c)
 		}
 	}()
+
+	c.traceInfo.tr = trace.New("Sent."+familyForMethod(method), method)
+	defer c.traceInfo.tr.Finish()
+	c.traceInfo.firstLine.client = true
+	if deadline, ok := ctx.Deadline(); ok {
+		c.traceInfo.firstLine.deadline = deadline.Sub(time.Now())
+	}
+	c.traceInfo.tr.LazyLog(&c.traceInfo.firstLine, false)
+	// TODO(dsymonds): Arrange for c.traceInfo.firstLine.remoteAddr to be set.
+	defer func() {
+		if err != nil {
+			c.traceInfo.tr.LazyLog(&lazySprintf{"%v", []interface{}{err}}, true)
+			c.traceInfo.tr.SetError()
+		}
+	}()
+
```

Comment:
On Sat, Jun 6, 2015 at 8:42 AM, David Symonds <notifications@github.com>
wrote:

> In call.go <https://github.com/grpc/grpc-go/pull/210#discussion_r31868998>
> :
>
> > +
> > +	c.traceInfo.tr = trace.New("Sent."+familyForMethod(method), method)
> > +	defer c.traceInfo.tr.Finish()
> > +	c.traceInfo.firstLine.client = true
> > +	if deadline, ok := ctx.Deadline(); ok {
> > +		c.traceInfo.firstLine.deadline = deadline.Sub(time.Now())
> > +	}
> > +	c.traceInfo.tr.LazyLog(&c.traceInfo.firstLine, false)
> > +	// TODO(dsymonds): Arrange for c.traceInfo.firstLine.remoteAddr to be set.
> > +	defer func() {
> > +		if err != nil {
> > +			c.traceInfo.tr.LazyLog(&lazySprintf{"%v", []interface{}{err}}, true)
> > +			c.traceInfo.tr.SetError()
> > +		}
> > +	}()
> > +
>
> What complexity are you thinking of? This change has 90% of the complexity
> it'll have for the client side, and the server side will be comparable.
>
This pull request probably won't add too much overhead (my only slight
concern is the string operations if method name is long). But if we enhance
this tracing later, it probably will run into some contention (e.g. update
some shared stats) and more expensive operations. So I prefer having a
switch at the 1st place.

> I'll take a look at benchmark_test.go.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-go/pull/210/files#r31868998>.
>